### ISSUE TITLE
Clang builds crash during startup on Ubuntu 16.04/18.04

### ISSUE THREAD

**Issue by troshko111** [envoyproxy/envoy#7027]
*Title*:
Envoy crashes when starting up with "attempt to free invalid pointer" in `tcmalloc` (`server.cc:229`).

*Description*:
When compiled with Clang (8.0.1) using the builder container and either `dbg` or `fastbuild` configuration (optimized builds do not crash but hang at the same place) Envoy segfaults at the startup.

When compiled without `tcmalloc` the crash occurs in the same place (free call).

*Repro steps*:
> 
1. Compile `285d04c3f7b64f9f0d78a2fd8f6a610625531c67` (or just `origin/master`) using the builder container with Clang toolchain in either `dbg` or `fastbuild`.
2. Run Envoy on Ubuntu 18.04 (upd: reproes on 16.04 as well).

*Config*:
> Linux troshko-thinkpad 5.0.7-100.fc28.x86_64 #1 SMP Mon Apr 8 16:46:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

> NAME="Ubuntu"
> VERSION="18.04.2 LTS (Bionic Beaver)"
> ID=ubuntu
> ID_LIKE=debian
> PRETTY_NAME="Ubuntu 18.04.2 LTS"
> VERSION_ID="18.04"
> HOME_URL="https://www.ubuntu.com/"
> SUPPORT_URL="https://help.ubuntu.com/"
> BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
> PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
> VERSION_CODENAME=bionic
> UBUNTU_CODENAME=bionic

Or 

> NAME="Ubuntu"
> VERSION="16.04.6 LTS (Xenial Xerus)"
> ID=ubuntu
> ID_LIKE=debian
> PRETTY_NAME="Ubuntu 16.04.6 LTS"
> VERSION_ID="16.04"
> HOME_URL="http://www.ubuntu.com/"
> SUPPORT_URL="http://help.ubuntu.com/"
> BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
> VERSION_CODENAME=xenial
> UBUNTU_CODENAME=xenial

```
{
  "admin": {
    "access_log_path": "/logs/envoy/admin_access.log",
    "address": {
      "socket_address": {
        "address": "127.0.0.1",
        "port_value": 1100
      }
    }
  },
  "static_resources": {
    "listeners": [
      {
        "name": "proxy_unencrypted_listener",
        "address": {
          "socket_address": {
            "address": "0.0.0.0",
            "port_value": 80
          }
        },
        "listener_filters": [
          {
            "name": "envoy.listener.tls_inspector",
            "config": {}
          }
        ],
        "filter_chains": [
          {
            "filters": [
              {
                "name": "envoy.http_connection_manager",
                "config": {
                  "stat_prefix": "ingress_local_http",
                  "codec_type": "AUTO",
                  "route_config": {
                    "name": "route_config",
                    "virtual_hosts": [
                      {
                        "name": "lo_vhost",
                        "domains": [
                          "*"
                        ],
                        "routes": [
                          {
                            "match": {
                              "prefix": "/"
                            },
                            "route": {
                              "cluster": "lo_svc_http",
                              "timeout": {
                                "seconds": 15
                              }
                            }
                          }
                        ]
                      }
                    ]
                  },
                  "http_filters": [
                    {
                      "name": "envoy.router"
                    }
                  ]
                }
              }
            ]
          }
        ]
      }
    ],
    "clusters": [
      {
        "name": "lo_svc_http",
        "connect_timeout": "2s",
        "type": "STATIC",
        "load_assignment": {
          "cluster_name": "lo_svc_http",
          "endpoints": [
            {
              "lb_endpoints": [
                {
                  "endpoint": {
                    "address": {
                      "socket_address": {
                        "address": "127.0.0.1",
                        "port_value": 8080
                      }
                    }
                  }
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

*Logs*:

```
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:202] initializing epoch 0 (hot restart version=11.104)
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:204] statically linked extensions:
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:206]   access_loggers: envoy.file_access_log,envoy.http_grpc_access_log
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:209]   filters.http: envoy.buffer,envoy.cors,envoy.csrf,envoy.ext_authz,envoy.fault,envoy.filters.http.eureka,envoy.filters.http.grpc_http1_reverse_bridge,envoy.filters.http.header_to_metadata,envoy.filters.http.jwt_authn,envoy.filters.http.rbac,envoy.filters.http.tap,envoy.grpc_http1_bridge,envoy.grpc_json_transcoder,envoy.grpc_web,envoy.gzip,envoy.health_check,envoy.http_dynamo_filter,envoy.ip_tagging,envoy.lua,envoy.rate_limit,envoy.router,envoy.squash
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:212]   filters.listener: envoy.listener.original_dst,envoy.listener.original_src,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:215]   filters.network: envoy.client_ssl_auth,envoy.echo,envoy.ext_authz,envoy.filters.network.dubbo_proxy,envoy.filters.network.mysql_proxy,envoy.filters.network.rbac,envoy.filters.network.sni_cluster,envoy.filters.network.thrift_proxy,envoy.filters.network.zookeeper_proxy,envoy.http_connection_manager,envoy.mongo_proxy,envoy.ratelimit,envoy.redis_proxy,envoy.tcp_proxy
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:217]   stat_sinks: envoy.dog_statsd,envoy.metrics_service,envoy.stat_sinks.atlas,envoy.stat_sinks.hystrix,envoy.statsd
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:219]   tracers: envoy.dynamic.ot,envoy.lightstep,envoy.tracers.datadog,envoy.zipkin
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:222]   transport_sockets.downstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2019-05-21 18:33:56.544][116][info][main] [external/envoy/source/server/server.cc:225]   transport_sockets.upstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2019-05-21 18:33:56.544][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Segmentation fault, suspect faulting address 0x7ffefffffff8
[2019-05-21 18:33:56.544][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-05-21 18:33:56.556][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #0: Envoy::SignalAction::sigHandler() [0x278d55c]
[2019-05-21 18:33:56.556][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #1: __restore_rt [0x7ffb01d1a890]
[2019-05-21 18:33:56.567][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #2: fmt::v5::basic_format_context<>::~basic_format_context() [0xfb8f29]
[2019-05-21 18:33:56.578][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #3: fmt::v5::format_handler<>::~format_handler() [0xfb8e09]
[2019-05-21 18:33:56.588][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #4: fmt::v5::vformat_to<>() [0x10728c4]
[2019-05-21 18:33:56.599][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #5: fmt::v5::vformat_to() [0x1072782]
[2019-05-21 18:33:56.610][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #6: fmt::v5::format_to<>() [0x109a66c]
[2019-05-21 18:33:56.621][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #7: spdlog::logger::log<>() [0x109a35b]
[2019-05-21 18:33:56.632][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #8: spdlog::logger::info<>() [0x153fc8a]
[2019-05-21 18:33:56.643][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #9: Envoy::Server::InstanceImpl::initialize() [0x1d871fc]
[2019-05-21 18:33:56.653][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #10: Envoy::Server::InstanceImpl::InstanceImpl() [0x1d8666d]
[2019-05-21 18:33:56.664][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #11: std::make_unique<>() [0x12df31b]
[2019-05-21 18:33:56.675][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #12: Envoy::MainCommonBase::MainCommonBase() [0x12dd0f0]
[2019-05-21 18:33:56.686][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #13: Envoy::MainCommon::MainCommon() [0x12ddc31]
[2019-05-21 18:33:56.697][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #14: std::make_unique<>() [0x12db150]
[2019-05-21 18:33:56.708][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #15: main [0x12dab7f]
[2019-05-21 18:33:56.708][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #16: [0x7ffb0159ab97]
```

*Call Stack*:

```
[2019-05-21 18:33:56.544][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Segmentation fault, suspect faulting address 0x7ffefffffff8
[2019-05-21 18:33:56.544][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-05-21 18:33:56.556][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #0: Envoy::SignalAction::sigHandler() [0x278d55c]
[2019-05-21 18:33:56.556][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #1: __restore_rt [0x7ffb01d1a890]
[2019-05-21 18:33:56.567][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #2: fmt::v5::basic_format_context<>::~basic_format_context() [0xfb8f29]
[2019-05-21 18:33:56.578][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #3: fmt::v5::format_handler<>::~format_handler() [0xfb8e09]
[2019-05-21 18:33:56.588][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #4: fmt::v5::vformat_to<>() [0x10728c4]
[2019-05-21 18:33:56.599][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #5: fmt::v5::vformat_to() [0x1072782]
[2019-05-21 18:33:56.610][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #6: fmt::v5::format_to<>() [0x109a66c]
[2019-05-21 18:33:56.621][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #7: spdlog::logger::log<>() [0x109a35b]
[2019-05-21 18:33:56.632][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #8: spdlog::logger::info<>() [0x153fc8a]
[2019-05-21 18:33:56.643][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #9: Envoy::Server::InstanceImpl::initialize() [0x1d871fc]
[2019-05-21 18:33:56.653][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #10: Envoy::Server::InstanceImpl::InstanceImpl() [0x1d8666d]
[2019-05-21 18:33:56.664][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #11: std::make_unique<>() [0x12df31b]
[2019-05-21 18:33:56.675][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #12: Envoy::MainCommonBase::MainCommonBase() [0x12dd0f0]
[2019-05-21 18:33:56.686][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #13: Envoy::MainCommon::MainCommon() [0x12ddc31]
[2019-05-21 18:33:56.697][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #14: std::make_unique<>() [0x12db150]
[2019-05-21 18:33:56.708][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #15: main [0x12dab7f]
[2019-05-21 18:33:56.708][116][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #16: [0x7ffb0159ab97]
```

*Call Stack (tcmalloc)*:
```
/source/.cache/bazel/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/netflix_envoy/external/com_github_gperftools_gperftools/src/tcmalloc.cc:332] Attempt to free invalid pointer 0x
[2019-05-21 19:54:15.405][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Aborted, suspect faulting address 0x11
[2019-05-21 19:54:15.405][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-05-21 19:54:15.417][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #0: Envoy::SignalAction::sigHandler() [0x27aa86c]
[2019-05-21 19:54:15.417][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #1: __restore_rt [0x7f6c9419e890]
[2019-05-21 19:54:15.428][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #2: (anonymous namespace)::InvalidFree() [0x2b0d5b3]
[2019-05-21 19:54:15.439][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #3: (anonymous namespace)::free_null_or_invalid() [0x2b0edb5]
[2019-05-21 19:54:15.450][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #4: tc_free [0x2bfa7e6]
[2019-05-21 19:54:15.461][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #5: fmt::v5::internal::arg_map<>::~arg_map() [0xfd608d]
[2019-05-21 19:54:15.472][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #6: fmt::v5::basic_format_context<>::~basic_format_context() [0xfd5f29]
[2019-05-21 19:54:15.483][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #7: fmt::v5::format_handler<>::~format_handler() [0xfd5e09]
[2019-05-21 19:54:15.494][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #8: fmt::v5::vformat_to<>() [0x108f8c4]
[2019-05-21 19:54:15.505][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #9: fmt::v5::vformat_to() [0x108f782]
[2019-05-21 19:54:15.516][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #10: fmt::v5::format_to<>() [0x10b766c]
[2019-05-21 19:54:15.526][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #11: spdlog::logger::log<>() [0x10b735b]
[2019-05-21 19:54:15.538][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #12: spdlog::logger::info<>() [0x155cc8a]
[2019-05-21 19:54:15.548][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #13: Envoy::Server::InstanceImpl::initialize() [0x1da41fc]
[2019-05-21 19:54:15.559][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #14: Envoy::Server::InstanceImpl::InstanceImpl() [0x1da366d]
[2019-05-21 19:54:15.570][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #15: std::make_unique<>() [0x12fc31b]
[2019-05-21 19:54:15.581][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #16: Envoy::MainCommonBase::MainCommonBase() [0x12fa0f0]
[2019-05-21 19:54:15.591][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #17: Envoy::MainCommon::MainCommon() [0x12fac31]
[2019-05-21 19:54:15.602][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #18: std::make_unique<>() [0x12f8150]
[2019-05-21 19:54:15.613][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #19: main [0x12f7b7f]
[2019-05-21 19:54:15.613][17][critical][backtrace] [bazel-out/k8-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #20: [0x7f6c93a1eb97]

```

I since compiled with and without tcmalloc and this does not seem to be related to tcmalloc at all, this is `gdb` starting at `server.cc:229`

> Run till exit from #0  0x000000000109a33c in spdlog::logger::log<char const*> (this=0x2d56900, lvl=spdlog::level::info, fmt=0x241a61 "[external/envoy/source/server/server.cc:231] buffer implementation: {}", args=@0x7fffffffd480: 0x26cd5f "old (libevent)") at external/spectator_spdlog/include/spdlog/details/logger_impl.h:66
> 
> Thread 1 "envoy" received signal SIGSEGV, Segmentation fault.
> 0x00007ffff70b298d in free () from /lib/x86_64-linux-gnu/libc.so.6

Both `tcmalloc` and non-`tcmalloc` versions crash when freeing the memory at the same place.

UPD: both Ubuntu LTS versions reproduce the issue.

**Comment by troshko111** [comment]
tl;dr I had another top level dependency compiled with a different version of `spdlog/fmtlib` and it seems like the wrong one was being called causing this crash.

Unrelated to Envoy, closing.
